File: SinglyLinkedList.java
public class SinglyLinkedList {
    private Node head;

    public SinglyLinkedList() { head = null; }

    void addFirst(String str) {
        Node first = new Node(str);
        first.setNext(head);
        head = first;
    }

    void addAfter(String insertAfter, String inputString) {
        Node curr = head;

        while (curr != null) {
            if (curr.getElement() == insertAfter) {
                Node newNode = new Node(inputString);
                newNode.setNext(curr.getNext());
                curr.setNext(newNode);
                break;
            }

            curr = curr.getNext();
        }
    }

    void addLast(String str) {
        if (head == null) {
            head = new Node(str);
        } else {
            Node curr = head;

            while (curr.getNext() != null) {
                curr = curr.getNext();
            }

            curr.setNext(new Node(str));
        }
    }

    Node removeFirst() {
        if (head != null) {
            Node delNode = head;
            head = head.getNext();
            return delNode;
        }

        return null;
    }

    Node removeAfter(String str) {
        Node curr = head;
        Node delNode = null;

        while (curr != null) {
            if (curr.getElement() == str && curr.getNext() != null) {
                delNode = curr.getNext();
                curr.setNext(delNode.getNext());
                break;
            }

            curr = curr.getNext();
        }

        return delNode;
    }

    Node removeLast() {
        Node curr = head;

        if (curr == null || curr.getNext() == null) {
            head = null;
            return curr;
        }

        Node nextNode = curr.getNext();

        while (curr.getNext() != null) {
            if (nextNode.getNext() == null) {
                curr.setNext(null);
            }

            curr = nextNode;
            nextNode = nextNode.getNext();
        }

        return curr;
    }

    @Override
    public String toString() {
        Node curr = head;

        String output = "\n[ ";

        while (curr.getNext() != null) {
            output += curr.getElement() + ", ";
            curr = curr.getNext();
        }

        output += curr.getElement() + " ]";

        return output;
    }
}

File: SinglyLinkedList.py
from Node import Node


class SinglyLinkedList:
    def __init__(self):
        self.head = None

    def addFirst(self, newNumber):
        newNode = Node(newNumber)
        newNode.next = self.head
        self.head = newNode

    def addAfter(self, afterNumber, newNumber):
        curr = self.head

        while curr is not None:
            if curr.data == afterNumber:
                newNode = Node(newNumber)
                newNode.next = curr.next
                curr.next = newNode
                break

            curr = curr.next

    def addLast(self, newNumber):
        if self.head is None:
            self.head = Node(newNumber)
        else:
            curr = self.head

            while curr.next is not None:
                curr = curr.next

            curr.next = Node(newNumber)

    def removeFirst(self):
        if self.head is not None:
            delNode = self.head
            self.head = self.head.next
            return delNode

        return None

    def removeAfter(self, removeAfterNumber):
        curr = self.head

        while curr is not None:
            if curr.data == removeAfterNumber and curr.next is not None:
                delNode = curr.next
                curr.next = delNode.next
                break

            curr = curr.next

        return delNode

    def removeLast(self):
        curr = self.head

        if curr is None or curr.next is None:
            self.head = None
            return curr

        nextNode = curr.next

        while curr.next is not None:
            if nextNode.next is None:
                curr.next = None

            curr = nextNode
            nextNode = nextNode.next

        return curr

    def printList(self):
        curr = self.head

        print()

        while curr:
            print(curr.data, end=' ')
            curr = curr.next

        print()
